lib/encoder.js

```diff
@@ -890,6 +907,95 @@ function defineInstanceMembers() {
     return Buffer.concat(parts, totalLength);
   };
 
+  this.decodeVector = function(buffer, params) {
+    const subtype = params["subtype"];
+    const dimensions = params["dimensions"];
+    const elemLength = 4; // TODO: figure this out based on the subtype
+    const expectedLength = buffer.length / elemLength;
+    if ((elemLength * dimensions) !== buffer.length) {
+      throw new TypeError(`Expected buffer of subtype ${subtype} with dimensions ${dimensions} to be of size ${expectedLength}, observed size ${buffer.length}`);
+    }
+    const rv = [];
+    let offset = 0;
+    for (let i = 0; i < dimensions; i++) {
+      offset = i * elemLength;
+      rv[i] = this.decode(buffer.slice(offset, offset + elemLength), subtype);
+    }
+    return new Float32Array(rv);
+  };
+
+  /**
+   * @param {CqlVector} value
+   * @param {Array} params
+   */
+  this.encodeVector = function(value, params) {
+
+    // Evaluate params to encodeVector(), returning the computed subtype
+    function evalParams() {
+
+      if (!(value instanceof Float32Array)) {
+        throw new TypeError("Driver only supports vectors of 4 byte floating point values");
+      }
+
+      // Perform client-side validation iff we were actually supplied with meaningful type info.  In practice
+      // this will only occur when using prepared statements.
+      if (params.hasOwnProperty("subtype") && params.hasOwnProperty("dimensions")) {
+
+        const subtype = params["subtype"];
+        const dimensions = params["dimensions"];
+        if (value.length !== dimensions) {
+          throw new TypeError(`Expected vector with ${dimensions} dimensions, observed size of ${value.length}`);
+        }
+        if (subtype.code !== dataTypes.float) {
+          throw new TypeError("Driver only supports vectors of 4 byte floating point values");
+        }
+        return subtype;
+      }
+
+      return { code: dataTypes.float };
+    }
+
+    if (!Encoder.isTypedArray(value)) {
+      throw new TypeError('Expected TypedArray subclass, obtained ' + util.inspect(value));
+    }
+    if (value.length === 0) {
+      throw new TypeError("Cannot encode empty array as vector");
+    }
+
+    const subtype = evalParams();
+
+    // TypedArrays are _not_ JS arrays so explicitly convert them here before trying to write them
+    // into a buffer
+    const elems = [];
+    for (const elem of value) {
+      elems.push(this.encode(elem, subtype));
+    }
+    return Buffer.concat(elems);
+  };
+
+  /**
+   * Extract the (typed) arguments from a vector type
+   *
+   * @param {String} typeName
+   * @param {String} stringToExclude Leading string indicating this is a vector type (to be excluded when eval'ing args)
+   * @param {Function} subtypeResolveFn Function used to resolve subtype type; varies depending on type naming convention
+   * @returns {Array}
```

Comment:
Is this return type doc-string correct? It looks like we're returning a dictionary (unsure of the exact syntax), similar question for second param of `encodeVector`

Reply:
Nope, that's out-of-date after some late code changes.  Good catch; I'll update this as well.